README.md
# NIFTY Cloud 向け knife プラグイン

## 概要

knife から NIFTY Cloud のインスタンスを扱うためのプラグインです。
knife-ec2 と同じように、インスタンスを作成した上で Chef のノードとして登録 (= ブートストラップ) したり、
インスタンスを削除したりできます。

## インストール

"chef" gem と ["ニフティクラウドSDK for Ruby" gem](http://cloud.nifty.com/api/sdk/) に依存しているので、
まだインストールしていない場合にはインストールしましょう。

	# "chef" gem のインストール
	gem install chef

	# "ニフティクラウドSDK for Ruby" のインストール
	wget http://cloud.nifty.com/api/sdk/NIFTY_Cloud_SDK_for_ruby.zip
	cd NIFTY_Cloud_SDK_for_ruby
	rake install

あとは gem install knife-nc を実行するだけです。

	gem install knife-nc

# 設定

NIFTY Cloud REST API と通信するため、knife コマンドにアクセスキー／シークレットキーを教えてやる必要があります。
knife.rb に設定するのが一番簡単でしょう。

	knife[:nc_access_key] = "Your NIFTY Cloud Access Key"
	knife[:nc_secret_key] = "Your NIFTY Cloud Access Key"

knife.rb をバージョン管理システムにコミットしている場合 (つまり誰でも閲覧可能になっている場合) には、
環境変数から読み込ませることも可能です。

	knife[:nc_access_key] = "#{ENV['NIFTY_CLOUD_ACCESS_KEY']}"
	knife[:nc_secret_key] = "#{ENV['NIFTY_CLOUD_SECRET_KEY']}"

アクセスキー／シークレットキーは knife サブコマンドの -A (--nc-access-key) と -K (--nc-secret-key) オプションで指定することもできます。

	# provision a new mini instance
	knife nc server create -r 'role[webserver]' -I 1 -T mini -A 'Your NIFTY Cloud Access Key' -K "Your NIFTY Cloud Secret Key"

さらに下記のオプションが knife.rb に設定可能です。

 * nc_instance_type
 * nc_image_id
 * nc_ssh_key_id
 * nc_bootstrap_version
 * nc_distro
 * nc_template_file
 * nc_user_data

## サブコマンド一覧

### knife nc server create

NIFTY Cloud 上にインスタンスを作成し SSH 経由で Chef のブートストラップを行います。「ブートストラップ」というのは Chef をインストールして Chef クライアントとして Chef サーバと通信できるようにすることです。利用するブートストラップ・スクリプトは指定したイメージ ID によって自動的に検出されますが、-d オプションや --template-file オプションで上書きすることもできます。インスタンス作成時、あとから knife コマンド経由で削除可能とするために、diableApiTermination を false に設定することに注意してください。

また、v0.0.1 からは独自オプションとして -R または --ssh-passphrase オプションで SSH パスフレーズが指定できるようになりました。

### knife nc server delete

設定済みの NIFTY Cloud アカウント内のインスタンスを指定して停止・削除します。注意：デフォルトではそのインスタンスに関するノード・クライアント情報を Chef サーバ上から削除しません。削除するためには --purge フラグをつけてください。

### knife nc server list

現在設定されている NIFTY Cloud アカウントが持つすべてのサーバの一覧を出力します。注意：アカウントに紐づいたすべてのインスタンスを表示するので、Chef サーバに管理されていないインスタンスが表示されることもあります。

### knife nc image list

knife-ec2 にはないオリジナルのコマンドです。現在のアカウントで利用可能なイメージの一覧を表示します。
knife nc server create コマンドの -I オプションに渡す Image ID を確認するために利用してください。

## NIFTY Cloud イメージ向けブートストラップ

knife-nc gem には NIFTY Cloud で提供されているイメージに最適化された 2 つのブートストラップ・スクリプト (nc-centos5-gems, nc-centos6-gems) が同梱されています。
knife nc server create を利用する際にはイメージ ID に応じて自動で利用すべきスクリプトの判別が行われます。
knife bootstrap からこのスクリプトを利用するには下記のように -d オプションで識別子を指定してください。

    knife bootstrap XXX.XXX.XXX.XXX -d nc-centos5-gems

なお、サポート対象のイメージ ID と各イメージ ID に対応するスクリプトの識別子は下記の通りです。

イメージ ID | OS・ディストリビューション | ブートストラップ・スクリプト | 同梱されている gem
------------|----------------------------|------------------------------|-------------------
1           | CentOS 5.3 32bit Plain     | nc-centos5-gems              | knife-nc
2           | CentOS 5.3 64bit Plain     | nc-centos5-gems              | knife-nc
6           | CentOS 5.3 32bit Server    | nc-centos5-gems              | knife-nc
7           | CentOS 5.3 64bit Server    | nc-centos5-gems              | knife-nc
13          | CentOS 5.6 64bit Plain     | nc-centos5-gems              | knife-nc
14          | CentOS 5.6 64bit Server    | nc-centos5-gems              | knife-nc
17          | Ubuntu 10.04 64bit Plain   | ubuntu10.04-gems             | chef
21          | CentOS 6.2 64bit Plain     | nc-centos6-gems              | knife-nc

## TODO

 * nc server create の --template-file オプション (ブートストラップ用テンプレート指定) の動作確認ができていない

## ライセンス

[knife-ec2](https://github.com/opscode/knife-ec2) をベースに NIFTY Cloud 向けに修正を加えたものです。
オリジナルに準じて Apache Lisence, Version 2.0 を適用します。


HISTORY.md
# 変更履歴

## 0.0.2

 * gem install knife-nc で nifty-cloud-sdk gem もインストールされるようになった
 * nc server delete コマンドの、サーバ停止しているのに停止しようとしてエラーになるバグを修正した
 * nc server delete コマンドの、disable_api_termination が true になっていると削除できないバグを修正した
 * nc server list コマンドの、サーバが 1 台も存在しないとエラー終了してしまうバグを修正した

## 0.0.1

version 0.0.0 がインストールされているとうまく動かないことがあるかもしれないので、gem uninstall knife-nc -v 0.0.0 したほうがいいかもです。

 * NIFTY Cloud 上の主要な 8 イメージに対応した
 * 起動時スクリプトに対応した
 * -R または --ssh-passphrase で SSH パスフレーズが指定できるようになった

## 0.0.0

 * 初期インポート


spec/spec_helper.rb
$:.unshift File.expand_path('../../lib', __FILE__)
require 'chef'
require 'chef/knife/nc_server_create'
require 'chef/knife/nc_server_delete'
require 'chef/knife/nc_server_list'
require 'chef/knife/nc_image_list'


lib/chef/knife/nc_base.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2011 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
$:.unshift(File.dirname(__FILE__) + '../../') unless $:.include?(File.dirname(__FILE__) + '../../')
require 'chef/knife'
require 'knife-nc/chef_extensions'
require 'knife-nc/nifty_cloud_sdk_extensions'

class Chef
  class Knife
    module NcBase

      # :nodoc:
      # Would prefer to do this in a rational way, but can't be done b/c of
      # Mixlib::CLI's design :(
      def self.included(includer)
        includer.class_eval do

          deps do
            require 'NIFTY'
            require 'readline'
            require 'chef/json_compat'
          end

          option :nc_access_key,
            :short => "-A ACCESS_KEY",
            :long => "--access-key KEY",
            :description => "Your NIFTY Cloud Access Key ID",
            :proc => Proc.new { |key| Chef::Config[:knife][:nc_access_key] = key }

          option :nc_secret_key,
            :short => "-K SECRET_KEY",
            :long => "--secret-key SECRET",
            :description => "Your NIFTY Cloud API Secret Access Key",
            :proc => Proc.new { |key| Chef::Config[:knife][:nc_secret_key] = key }
        end
      end

      def connection
        @connection ||= begin
          connection = NIFTY::Cloud::Base.new(
            :access_key => Chef::Config[:knife][:nc_access_key],
            :secret_key => Chef::Config[:knife][:nc_secret_key]
        )
        end
      end

      def locate_config_value(key)
        key = key.to_sym
        Chef::Config[:knife][key] || config[key]
      end

      def msg_pair(label, value, color=:cyan)
        if value && !value.to_s.empty?
          puts "#{ui.color(label, color)}: #{value}"
        end
      end

      def validate!(keys=[:nc_access_key, :nc_secret_key])
        errors = []

        keys.each do |k|
          pretty_key = k.to_s.gsub(/_/, ' ').gsub(/\w+/){ |w| (w =~ /(ssh)|(nc)|(id)/i) ? w.upcase  : w.capitalize }
          if Chef::Config[:knife][k].nil?
            errors << "You did not provide a valid '#{pretty_key}' value."
          end
        end

        if errors.each{|e| ui.error(e)}.any?
          exit 1
        end
      end

    end
  end
end




lib/chef/knife/nc_image_list.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2012 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/knife/nc_base'

class Chef
  class Knife
    class NcImageList < Knife

      include Knife::NcBase

      banner "knife nc image list (options)"

      def run
        $stdout.sync = true

        validate!

        image_list = [
          ui.color('Image ID', :bold),
          ui.color('Name', :bold),
          ui.color('Owner', :bold),
          ui.color('State', :bold)
        ]
        connection.describe_images.imagesSet.item.each do |image|
          image_list << image.imageId
          image_list << image.name
          image_list << "#{image.imageOwnerId} (#{image.imageOwnerAlias})"
          image_list << "#{image.imageState} (#{image.isPublic ? 'public' : 'private'})"
        end
        puts ui.list(image_list, :columns_across, 4)

      end
    end
  end
end




lib/knife-nc/chef_extensions.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2011 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
require 'chef/knife'
require 'chef/knife/ssh'
require 'chef/knife/bootstrap'

class Chef::Knife
  class Ssh
    alias original_session_from_list session_from_list

    option :ssh_passphrase,
      :long => "--ssh-passphrase PASSPHRASE",
      :description => "Your SSH Passphrase",
      :proc => Proc.new { |key| Chef::Config[:knife][:ssh_passphrase] = key }

    def session_from_list(list)
      session = original_session_from_list(list)
      session.servers.each do |server|
        server.options[:passphrase] = config[:ssh_passphrase]
      end
      session
    end
  end

  class Bootstrap
    alias original_knife_ssh knife_ssh

    option :ssh_passphrase,
      :long => "--ssh-passphrase PASSPHRASE",
      :description => "Your SSH Passphrase",
      :proc => Proc.new { |key| Chef::Config[:knife][:ssh_passphrase] = key }

    def knife_ssh
      ssh = original_knife_ssh
      ssh.config[:ssh_passphrase] = config[:ssh_passphrase]
      ssh
    end
  end
end


lib/chef/knife/nc_server_list.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2012 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/knife/nc_base'

class Chef
  class Knife
    class NcServerList < Knife

      include Knife::NcBase

      banner "knife nc server list (options)"

      def run
        $stdout.sync = true

        validate!

        server_list = [
          ui.color('Instance ID', :bold),
          ui.color('Public IP', :bold),
          ui.color('Private IP', :bold),
          ui.color('Flavor', :bold),
          ui.color('Image', :bold),
          ui.color('SSH Key', :bold),
          ui.color('Security Group', :bold),
          ui.color('State', :bold)
        ]
        set = connection.describe_instances.reservationSet
        if set
          set.item.each do |instance|
            server = instance.instancesSet.item.first
            group = instance.groupSet
            server_list << server.instanceId
            server_list << server.ipAddress.to_s
            server_list << server.privateIpAddress.to_s
            server_list << server.instanceType
            server_list << server.imageId
            server_list << server.keyName
            server_list << (group ? group.item.first.groupId : '')
            server_list << begin
              state = server.instanceState.name
              case state
              when 'sotopped', 'warning', 'waiting', 'creating', 'suspending', 'uploading', 'import_error'
                ui.color(state, :red)
              when 'pending'
                ui.color(state, :yellow)
              else
                ui.color(state, :green)
              end
            end
          end
        end
        puts ui.list(server_list, :columns_across, 8)

      end
    end
  end
end




lib/chef/knife/nc_server_delete.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2012 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/knife/nc_base'

# These two are needed for the '--purge' deletion case
require 'chef/node'
require 'chef/api_client'

class Chef
  class Knife
    class NcServerDelete < Knife

      include Knife::NcBase

      banner "knife nc server delete INSTANCE_ID_1 [INSTANCE_ID_2 ...] (options)"

      option :purge,
        :short => "-P",
        :long => "--purge",
        :boolean => true,
        :default => false,
        :description => "Destroy corresponding node and client on the Chef Server, in addition to destroying the NIFTY Cloud node itself.  Assumes node and client have the same name as the server (if not, add the '--node-name' option)."

      option :chef_node_name,
        :short => "-N NAME",
        :long => "--node-name NAME",
        :description => "The name of the node and client to delete, if it differs from the server name.  Only has meaning when used with the '--purge' option."

      option :force,
        :long => "--force",
        :default => false,
        :description => "Force to delete server by changing the server's disableApiTermination attribute to false."

      # Extracted from Chef::Knife.delete_object, because it has a
      # confirmation step built in... By specifying the '--purge'
      # flag (and also explicitly confirming the server destruction!)
      # the user is already making their intent known.  It is not
      # necessary to make them confirm two more times.
      def destroy_item(klass, name, type_name)
        begin
          object = klass.load(name)
          object.destroy
          ui.warn("Deleted #{type_name} #{name}")
        rescue Net::HTTPServerException
          ui.warn("Could not find a #{type_name} named #{name} to delete!")
        end
      end

      def run

        validate!

        @name_args.each do |instance_id|

          begin
            server = connection.describe_instances(:instance_id => instance_id).reservationSet.item.first.instancesSet.item.first

            msg_pair("Instance ID", server.instanceId)
            msg_pair("Instance Type", server.instanceType)
            msg_pair("Image ID", server.imageId)
            #msg_pair("Security Groups", server.groups.join(", "))
            msg_pair("SSH Key", server.keyName)
            msg_pair("Root Device Type", server.root_device_type)
            msg_pair("Public IP Address", server.ipAddress)
            msg_pair("Private IP Address", server.privateIpAddress)

            puts "\n"
            confirm("Do you really want to delete this server")

            if server.instanceState.name != 'stopped'
              print "\n#{ui.color("Waiting for server to shutdown", :magenta)}"
              connection.stop_instances(:instance_id => instance_id, :force => true)
              while server.instanceState.name != 'stopped'
                print "."
                server = connection.describe_instances(:instance_id => instance_id).reservationSet.item.first.instancesSet.item.first
                sleep 5
              end
              puts("done\n")
            end

            attribute = connection.describe_instance_attribute(:instance_id => instance_id, :attribute => 'disableApiTermination')
            if attribute.disableApiTermination.value != 'false'
              if config[:force] == false
                ui.error("Server's 'disableApiTermination' attribute is true. Use --force option to delete server.")
                exit 1
              else
                print "\n#{ui.color("Enabling API termination for server", :magenta)}"
                connection.modify_instance_attribute(:instance_id => instance_id, :attribute => 'disableApiTermination', :value => 'false')
                while attribute.disableApiTermination.value != 'false'
                  print "."
                  attribute = connection.describe_instance_attribute(:instance_id => instance_id, :attribute => 'disableApiTermination')
                  sleep 5
                end
                puts("done\n")
              end
            end

            connection.terminate_instances(:instance_id => instance_id)

            ui.warn("Deleted server #{server.id}")

            if config[:purge]
              thing_to_delete = config[:chef_node_name] || instance_id
              destroy_item(Chef::Node, thing_to_delete, "node")
              destroy_item(Chef::ApiClient, thing_to_delete, "client")
            else
              ui.warn("Corresponding node and client for the #{instance_id} server were not deleted and remain registered with the Chef Server")
            end

          rescue NoMethodError
            ui.error("Could not locate server '#{instance_id}'.  Please verify it was provisioned in the '#{locate_config_value(:region)}' region.")
          end
        end
      end

    end
  end
end


lib/knife-nc/nifty_cloud_sdk_extensions.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2011 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
require 'NIFTY'

class NIFTY::Cloud::Base
  alias original_run_instances run_instances
  alias original_response_generator response_generator
  def run_instances(options)
    options[:password] = 'ignoreme' unless options[:password]
    @user_data = options[:user_data]
    original_run_instances(options)
  end

  def response_generator(params)
    params.delete('Password') if params['Password'] == 'ignoreme'
    if @user_data
      params['UserData'] = extract_user_data(:user_data => @user_data, :base64_encoded => true)
      params['UserData.Encoding'] = 'base64'
    end
    original_response_generator(params)
  end
end


lib/chef/knife/nc_server_create.rb
#
# Author:: tily (<tidnlyam@gmail.com>)
# Copyright:: Copyright (c) 2012 tily
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/knife/nc_base'

class Chef
  class Knife
    class NcServerCreate < Knife

      include Knife::NcBase

      deps do
        require 'NIFTY'
        require 'readline'
        require 'chef/json_compat'
        require 'chef/knife/bootstrap'
        Chef::Knife::Bootstrap.load_deps
      end

      banner "knife nc server create INSTANCE_ID (options)"

      attr_accessor :initial_sleep_delay

      option :nc_instance_type,
        :short => "-T INSTANCE_TYPE",
        :long => "--instance-type INSTANCE_TYPE",
        :description => "The instance type of server (small, medium, etc)",
        :proc => Proc.new { |t| Chef::Config[:knife][:nc_instance_type] = t },
        :default => "small"

      option :nc_image_id,
        :short => "-I IMAGE_ID",
        :long => "--image IMAGE_ID",
        :description => "The Image ID for the server",
        :proc => Proc.new { |i| Chef::Config[:knife][:nc_image_id] = i }

      option :security_group,
        :short => "-G GROUP",
        :long => "--group GROUP",
        :description => "The security group for this server"

      option :chef_node_name,
        :short => "-N NAME",
        :long => "--node-name NAME",
        :description => "The Chef node name for your new node"

      option :nc_ssh_key_id,
        :short => "-S KEY",
        :long => "--ssh-key KEY",
        :description => "The NIFTY Cloud SSH key id",
        :proc => Proc.new { |key| Chef::Config[:knife][:nc_ssh_key_id] = key }

      option :ssh_user,
        :short => "-x USERNAME",
        :long => "--ssh-user USERNAME",
        :description => "The ssh username",
        :default => "root"

      option :ssh_password,
        :short => "-P PASSWORD",
        :long => "--ssh-password PASSWORD",
        :description => "The ssh password"

      option :ssh_passphrase,
        :short => "-R PASSPHRASE",
        :long => "--ssh-passphrase PASSPHRASE",
        :description => "The ssh passphrase",
        :proc => Proc.new { |key| Chef::Config[:knife][:ssh_passphrase] = key }

      option :ssh_locally,
        :short => "-L",
        :long => "--ssh-locally",
        :description => "bootstrap via ssh using private IP address"

      option :identity_file,
        :short => "-i IDENTITY_FILE",
        :long => "--identity-file IDENTITY_FILE",
        :description => "The SSH identity file used for authentication"

      option :prerelease,
        :long => "--prerelease",
        :description => "Install the pre-release chef gems"

      option :nc_bootstrap_version,
        :long => "--bootstrap-version VERSION",
        :description => "The version of Chef to install",
        :proc => Proc.new { |v| Chef::Config[:knife][:nc_bootstrap_version] = v }

      option :nc_distro,
        :short => "-d DISTRO",
        :long => "--distro DISTRO",
        :description => "Bootstrap a distro using a template",
        :proc => Proc.new { |d| Chef::Config[:knife][:nc_distro] = d }

      option :nc_template_file,
        :long => "--template-file TEMPLATE",
        :description => "Full path to location of template to use",
        :proc => Proc.new { |t| Chef::Config[:knife][:nc_template_file] = t },
        :default => false

      option :run_list,
        :short => "-r RUN_LIST",
        :long => "--run-list RUN_LIST",
        :description => "Comma separated list of roles/recipes to apply",
        :proc => lambda { |o| o.split(/[\s,]+/) },
        :default => []

      option :no_host_key_verify,
        :long => "--no-host-key-verify",
        :description => "Disable host key verification",
        :boolean => true,
        :default => false

      option :nc_user_data,
        :long => "--user-data USER_DATA_FILE",
        :short => "-U USER_DATA_FILE",
        :description => "The NIFTY Cloud User Data file to provision the instance with",
        :proc => Proc.new { |m| Chef::Config[:knife][:nc_user_data] = m },
        :default => nil

      def tcp_test_ssh(hostname)
        tcp_socket = TCPSocket.new(hostname, 22)
        readable = IO.select([tcp_socket], nil, nil, 5)
        if readable
          Chef::Log.debug("sshd accepting connections on #{hostname}, banner is #{tcp_socket.gets}")
          yield
          true
        else
          false
        end
      rescue SocketError
        sleep 2
        false
      rescue Errno::ETIMEDOUT
        false
      rescue Errno::EPERM
        false
      rescue Errno::ECONNREFUSED
        sleep 2
        false
      rescue Errno::EHOSTUNREACH
        sleep 2
        false
      ensure
        tcp_socket && tcp_socket.close
      end

      def run
        $stdout.sync = true
        NIFTY::LOG.level = Logger::DEBUG

        validate!

        msg_pair("Instance ID", @name_args.first)
        msg_pair("Instance Type", locate_config_value(:nc_instance_type))
        msg_pair("Image", locate_config_value(:nc_image_id))
        msg_pair("SSH Key", locate_config_value(:nc_ssh_key_id))
        msg_pair("Security Group", config[:security_group])

        puts "\n"
        confirm("Do you really want to create this server")

        server = connection.run_instances(create_server_def).instancesSet.item.first

        msg_pair("Instance ID", server.instanceId)
        msg_pair("Instance Type", server.instanceType)
        msg_pair("Image", server.imageId)
        msg_pair("SSH Key", server.keyName)

        print "\n#{ui.color("Waiting for server", :magenta)}"

        # wait for it to be ready to do stuff
        while server.instanceState.name != 'running'
          print "."
          server = connection.describe_instances(:instance_id => server.instanceId).reservationSet.item.first.instancesSet.item.first
          sleep 5
        end

        puts("done\n")
        
        msg_pair("Public IP Address", server.ipAddress)
        msg_pair("Private IP Address", server.privateIpAddress)

        ssh_ip_address = config[:ssh_locally] ? server.privateIpAddress : server.ipAddress
        print "\n#{ui.color("Waiting for sshd", :magenta)} (using #{ssh_ip_address})"
        
        print(".") until tcp_test_ssh(ssh_ip_address) {
          sleep @initial_sleep_delay ||= 10
          puts("done")
        }

        bootstrap_for_node(server, ssh_ip_address).run

        puts "\n"
        msg_pair("Instance ID", server.instanceId)
        msg_pair("Instance Type", server.instanceType)
        msg_pair("Image ID", server.imageId)
        msg_pair("Security Group", config[:security_group])
        msg_pair("SSH Key", server.keyName)
        msg_pair("Public IP Address", server.ipAddress)
        msg_pair("Private IP Address", server.privateIpAddress)
        msg_pair("Environment", config[:environment] || '_default')
        msg_pair("Run List", config[:run_list].join(', '))
      end

      def bootstrap_for_node(server,fqdn)
        bootstrap = Chef::Knife::Bootstrap.new
        bootstrap.name_args = [fqdn]
        bootstrap.config[:run_list] = config[:run_list]
        bootstrap.config[:ssh_user] = config[:ssh_user]
        bootstrap.config[:ssh_password] = config[:ssh_password]
        bootstrap.config[:ssh_passphrase] = config[:ssh_passphrase]
        bootstrap.config[:identity_file] = config[:identity_file]
        bootstrap.config[:chef_node_name] = config[:chef_node_name] || server.instanceId
        bootstrap.config[:prerelease] = config[:prerelease]
        bootstrap.config[:bootstrap_version] = locate_config_value(:nc_bootstrap_version)

        image_id = locate_config_value(:nc_image_id)
	if locate_config_value(:nc_distro)
          bootstrap.config[:distro] = locate_config_value(:nc_distro)
        elsif %w(1 2 6 7 13 14).include?(image_id)
          bootstrap.config[:distro] = 'nc-centos5-gems'
        elsif image_id == '17'
          bootstrap.config[:distro] = 'ubuntu10.04-gems'
        elsif image_id == '21'
          bootstrap.config[:distro] = 'nc-centos6-gems'
	else
          bootstrap.config[:distro] = 'nc-centos5-gems'
        end

        bootstrap.config[:use_sudo] = true unless config[:ssh_user] == 'root'
        bootstrap.config[:template_file] = locate_config_value(:nc_template_file)
        bootstrap.config[:environment] = config[:environment]
        bootstrap
      end

      def image
        @image ||= connection.describe_images(:image_id => locate_config_value(:nc_image_id)).imagesSet.item.first
      end

      def validate!
        super([:nc_image_id, :nc_ssh_key_id, :nc_access_key, :nc_secret_key])

        errors = []

        if @name_args.empty?
          errors << "You have not provided a valid instance ID."
        end

        if config[:security_group].nil?
          errors << "You have not provided a valid Security Group."
        end

        [:identity_file, :nc_template_file, :nc_user_data].each do |x|
          if (path = locate_config_value(x)) && !File.exists?(path)
            errors << "File does not exist: #{path}"
          end
        end

        if errors.each{|e| ui.error(e)}.any?
          exit 1
        end
      end
      
      def create_server_def
        server_def = {
          :instance_id => @name_args.first,
          :image_id => locate_config_value(:nc_image_id),
          :instance_type => locate_config_value(:nc_instance_type),
          :security_group => config[:security_group],
          :key_name => Chef::Config[:knife][:nc_ssh_key_id],
          :disable_api_termination => false # for 'knife nc server delete' command
        }

        if Chef::Config[:knife][:nc_user_data]
          begin
            server_def.merge!(:user_data => File.read(locate_config_value(:nc_user_data)))
          rescue
            ui.warn("Cannot read #{Chef::Config[:knife][:nc_user_data]}: #{$!.inspect}. Ignoring option.")
          end
        end

        server_def
      end
    end
  end
end
